File: sushiwindow/sushiwindow/Window.swift
//
//  Window.swift
//  sushiwindow
//
//  Created by Michael Ong on 22/6/17.
//  Copyright © 2017 Michael Ong. All rights reserved.
//

import Foundation

import sushicore
import sushigraphics

public class Window: WindowDriverDelegate {

    private var     _location           : Rectangle     = .zero

    public var      location            : Rectangle     {

        get     { return _location }
        set(d)  {

            if d == Rectangle.undefined { _location = driver.undefinedRect() }
            else                        { _location = d }

            driver.process(message: .sizeChanged,    data: WindowLocationDpi(location: location, dpi: dpi))

            resizeSurface()
        }
    }


    private var     _dpi                : Double        = 0

    public var      dpi                 : Double        {

        get     { return _dpi }
        set(d)  {

            _dpi = d

            driver.process(message: .sizeChanged,    data: WindowLocationDpi(location: location, dpi: dpi))

            resizeSurface()
        }
    }


    private var     _title              : String        = ""

    public var      title               : String        {

        get     { return _title }
        set(d)  {

            _title = d

            driver.process(message: .titleChanged,   data: title)
        }
    }


    private var     _state              : Set<WindowState> = [ ]

    public var      state               : Set<WindowState> {

        get { return _state }
        set(d) {

            _state = d

            driver.process(message: .stateChanged,  data: _state)

            logic.stateChanged(to: _state)
        }
    }


    public var      driver              : WindowDriver

    public private(set)
    var             graphicsSurface     : GraphicsSurface!


    public internal(set)
    var             logic               : WindowLogic


    public var      backingWindow       : Any!


    public init                         (with driver: WindowDriver, logic: WindowLogic, _ graphics: GraphicsDevice) {

        assert(driver.delegate == nil)

        self.logic              = logic
        self.driver             = driver

        self.logic.window       = self
        self.driver.delegate    = self

        self.backingWindow      = driver.initialise()
        self.graphicsSurface    = try? graphics.createSurface(with: _location.size, _dpi, to: backingWindow)
    }

    public func     send                (message: WindowDriverMessage, data: Any) {

        switch message {

        case .sizeChanged:

            guard let theData = data as? WindowLocationDpi  else { FastFail.instance.crash(with: .unexpectedPayload) }

            _location   = theData.location
            _dpi        = theData.dpi

        case .stateChanged:

            guard let theData = data as? Set<WindowState>   else { FastFail.instance.crash(with: .unexpectedPayload) }

            _state      = theData

        case .titleChanged:

            guard let theData = data as? String             else { FastFail.instance.crash(with: .unexpectedPayload) }

            _title      = theData
        }
    }


    private func    resizeSurface       () {

        try? self.graphicsSurface.resize(to: _location.size, _dpi)
    }
}

extension Rectangle {

    public static var undefined: Rectangle { return .init(x: 0, y: 0, width: -1, height: -1) }
}

File: sushicpp/sushiwindow/sushiwindow/impl/window.cpp
// window.cpp
// sushiwindow - sushicpp
// 
// Created: 13:20 08/19/2017
// Author: Michael Ong

#include "../window.hpp"
#include "../windowLogic.hpp"
#include "../windowEncapsulationIDs.hpp"

#include <sushicore/FastFail.h>

#include <sushiapplication/application.hpp>

#include <sushigraphics/graphicsDevice.hpp>
#include <sushigraphics/graphicsSurface.hpp>

using namespace std;

using namespace sushi::core;
using namespace sushi::application;
using namespace sushi::window;

using namespace sushi::drivers;

Rectangle                   Window::undefinedRectangle  = Rectangle { glm::vec2 { 0, 0 }, glm::vec2 { -1, -1 } };


WindowEncapsulatedMessage*  Window::encapsulate         (unsigned int dataID, void* data) {

    auto message = new WindowEncapsulatedMessage();

    message->data = data;
    message->dataID = dataID;

    return message;
}


Window::Window                          (WindowDriver* driver, WindowLogic* logic):

CoreObject              (),
WindowDriverDelegate    (),

_dpi                    (0),
_title                  (u"Title Goes Here!"),
_windowState            (),

_backingWindow          (nullptr),

_windowDriver           (driver),
_graphicsSurface        (nullptr),
_inputDispatcher        (nullptr),

_windowLogic            (logic) {

    assert(_windowDriver->delegate == nullptr);

    _windowDriver   ->delegate  = this;
    _windowLogic    ->window    = this;

    _backingWindow              = _windowDriver->initalise();

    _inputDispatcher            = nullptr;
    _graphicsSurface            = Application::instance->getGraphicsDevice()->createSurface({ 1, 1 }, 1, _backingWindow);
}

Window::Window                          (const Window&& move):

_dpi                    (move._dpi),
_title                  (std::move(move._title)),
_windowState            (std::move(move._windowState)),

_backingWindow          (move._backingWindow),

_windowDriver           (move._windowDriver),
_graphicsSurface        (move._graphicsSurface),
_inputDispatcher        (move._inputDispatcher),

_windowLogic            (move._windowLogic) {

    _windowDriver->delegate = this;
    _windowLogic->window = this;
}

Window::~Window                         () {

    delete _windowLogic;
    delete _windowDriver;
}


void        Window::setLocation        (const Rectangle& value) {

    if (_location != value) {

        auto finalValue = value;

        if (finalValue == undefinedRectangle) {

            finalValue = _windowDriver->getUndefinedRect();
        }

        _location = finalValue;

        _windowDriver->process(WindowDriverMessages::sizeChanged, nullptr);
    }
}

void        Window::setDPI             (const float& value) {

    if (_dpi != value) {
        
        _dpi = value;

        _windowDriver->process(WindowDriverMessages::sizeChanged, nullptr);
    }
}

void        Window::setTitle           (const u16string& value) {

    if (_title != value) {

        _title = value;

        _windowDriver->process(WindowDriverMessages::titleChanged, nullptr);
    }
}

void        Window::setWindowState     (const set<WindowStates> &value) {

    if (_windowState != value) {

        _windowState = value;

        _windowDriver->process(WindowDriverMessages::stateChanged, nullptr);
    }
}

void        Window::send               (WindowDriverMessages message, WindowEncapsulatedMessage* data) {

    switch (message) {

        case WindowDriverMessages::sizeChanged: {

            auto resizeNeeded = false;

            if (data->dataID == WindowEncapsulationIDs::valueFloat) {

                const auto newValue = float(*reinterpret_cast<float*>(data->data));

                if (newValue != _dpi) {

                    _dpi = newValue;
                    resizeNeeded = true;
                }

            } else if (data->dataID == WindowEncapsulationIDs::valueRectangle) {

                const auto newValue = Rectangle(*reinterpret_cast<Rectangle*>(data->data));

                if (newValue != _location) {

                    _location = newValue;
                    resizeNeeded = true;
                }

            } else {

                FastFail::instance->crash(failureTypes::unexpectedPayload("(Window::send): Expecting a float or a Rectangle but got something else."));
            }

            if (resizeNeeded) {

                _graphicsSurface->resize(_location.size, _dpi);
            }

        }   break;
        case WindowDriverMessages::titleChanged:

            if (data->dataID == WindowEncapsulationIDs::valueString) {

                _title = u16string(*reinterpret_cast<u16string*>(data->data));

            } else {

                FastFail::instance->crash(failureTypes::unexpectedPayload("(Window::send): Expecting a u16string but got something else."));
            }

            break;
        case WindowDriverMessages::stateChanged:

            if (data->dataID == WindowEncapsulationIDs::valueSet) {

                _windowState = set<WindowStates>(*reinterpret_cast<set<WindowStates>*>(data->data));

            } else {

                FastFail::instance->crash(failureTypes::unexpectedPayload("(Window::send): Expecting a set<WindowStates> but got something else."));
            }

            break;
    }
}
